src/lem/circuit.rs

```diff
@@ -0,0 +1,899 @@
+                    z_ptr.tag.to_field(),
+                )?);
+
+                component_idx += 1;
+
+                // allocate pointer hash
+                preallocated_preimg.push(Self::allocate_preimg_component_for_slot(
+                    cs,
+                    &slot,
+                    component_idx,
+                    z_ptr.hash,
+                )?);
+
+                component_idx += 1;
+            }
+
+            // Allocate the image by calling the arithmetic function according
+            // to the slot type
+            let preallocated_img =
+                Self::allocate_img_for_slot(cs, &slot, preallocated_preimg.clone(), store)?;
+
+            preallocations.push((preallocated_preimg, preallocated_img));
+        }
+
+        // Then we do the same with dummies for the remaining slots
+        for slot_idx in preallocations.len()..num_slots {
+            let slot = Slot {
+                idx: slot_idx,
+                typ: slot_type,
+            };
+            let mut preallocated_preimg = vec![];
+            for component_idx in 0..slot_type.preimg_size() {
+                preallocated_preimg.push(Self::allocate_preimg_component_for_slot(
+                    cs,
+                    &slot,
+                    component_idx,
+                    F::ZERO,
+                )?);
+            }
+            let preallocated_img =
+                Self::allocate_img_for_slot(cs, &slot, preallocated_preimg.clone(), store)?;
+
+            preallocations.push((preallocated_preimg, preallocated_img));
+        }
+        Ok(preallocations)
+    }
+
+    /// Create R1CS constraints for LEM given an evaluation frame. This function
+    /// implements the STEP 3 mentioned above.
+    ///
+    /// Notably, we recursively pass on a `concrete_path: Boolean` variable, which
+    /// encodes whether we are on a *concrete* or *virtual* path. A virtual path
+    /// is one that wasn't taken during interpretation and thus its frame pointers
+    /// weren't bound. A concrete path means that interpretation went down that
+    /// road and the frame data should be complete for the variables on that path.
+    /// For virtual paths we need to create dummy bindings for the meta pointers
+    /// and relax the implications with false premises. The premise is precicely
+    /// `concrete_path`.
+    ///
+    /// Regarding the slot optimizations, STEP 3 uses information gathered during
+    /// STEPs 1 and 2. So we proceed by first allocating preimages and images for
+    /// each slot and then, as we traverse the LEM, we add constraints to make sure
+    /// that the witness satisfies the arithmetic equations for the corresponding
+    /// slots.
+    pub fn synthesize<F: LurkField, CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        store: &mut Store<F>,
+        slots_count: &SlotsCounter,
+        frame: &Frame<F>,
+    ) -> Result<()> {
+        let alloc_manager = AllocationManager::default();
```

Comment:
I think the `AllocationManager` (can this be an `allocator`?) Should be shared between all frames of a given circuit (when rc != 1). In that case, we shouldn't produce a new one per synthesized frame, right?

This only matters if you are using the allocator to deduplicate 'constants', like pointers to common symbols, etc. However, my memory is that you are. Right?

Reply:
@gabriel-barrett this is why the `AllocationManager` needs to be passed to `synthesize`

Reply:
This will make more sense once I do the change I mentioned here: https://github.com/lurk-lab/lurk-rs/pull/427#discussion_r1243074120

Reply:
I called it `GlobalAllocator` to explicit the link with `GlobalAllocations`

Reply:
I'm gonna mark this as resolved, because Multiframe will be left for some other PR. Right now `AllocationManager` is not an argument to `synthesize` in Frame because it is dangerous if you use it manually. They should only be shared when building the circuit from a MultiFrame